### PR TITLE
Sygus enumerators to conjecture

### PR DESCRIPTION
**Pull request by ajreynol**
This commit maps enumerated terms to the Sygus conjectures they are associated with. This is partial work towards #1210.

It moves PBE utilities from TermDbSygus so that they are kept local to CegConjecturePbe, the utility for doing PBE-specific optimizations.

Next steps:
- Further split and document the SyGuS term database (#1235).
- Migrate PBE-specific code to CegConjecture and generalize to other conjecture-specific symmetry breaking techniques.

### CODE REVIEW

File: src/theory/quantifiers/ce_guided_pbe.h [hunk 1]
```diff
@@ -64,7 +87,42 @@ class CegConjecturePbe {
...
+  void getExample( Node e, unsigned i, std::vector< Node >& ex );
+  /** get the output value of the i^th I/O example for enumerator e */
+  Node getExampleOut( Node e, unsigned i );
+  int getExampleId( Node n );
+  /** add the search val, returns an equivalent value (possibly the same) */
+  Node addSearchVal( TypeNode tn, Node e, Node bvr );
+  /** evaluate builtin */
+  Node evaluateBuiltin( TypeNode tn, Node bn, Node e, unsigned i );
+  
+private:
```
Comment (mpreiner) [inline src/theory/quantifiers/ce_guided_pbe.h]: Can you get rid of the mixed private/public sections?

File: src/theory/quantifiers/ce_guided_pbe.h [hunk 2]
```diff
@@ -64,7 +87,42 @@ class CegConjecturePbe {
...
+  /** this class is an index of candidate solutions for PBE synthesis */
+  class PbeTrie {
+  private:
```
Comment (mpreiner) [inline src/theory/quantifiers/ce_guided_pbe.h]: `private:` after `public:`

File: src/theory/quantifiers/term_database_sygus.cpp [hunk 3]
```diff
@@ -22,6 +22,7 @@
 #include "theory/quantifiers/theory_quantifiers.h"
 #include "theory/quantifiers/term_database.h"
 #include "theory/quantifiers/term_util.h"
+#include "theory/quantifiers/ce_guided_conjecture.h"
 #include "theory/quantifiers_engine.h"
```
Comment (mpreiner) [inline src/theory/quantifiers/term_database_sygus.cpp]: Alphabetize

File: src/theory/quantifiers/term_database_sygus.h [hunk 4]
```diff
@@ -23,6 +23,8 @@ namespace CVC4 {
 namespace theory {
 namespace quantifiers {
 
+class CegConjecture;
```
Comment (mpreiner) [inline src/theory/quantifiers/term_database_sygus.h]: Use `#include theory/quantifieres/ce_guided_conjecture.h`

File: src/theory/quantifiers/ce_guided_pbe.cpp [hunk 5]
```diff
@@ -199,26 +199,129 @@ void CegConjecturePbe::initialize( Node n, std::vector< Node >& candidates, std:
...
+
+Node CegConjecturePbe::getExampleOut( Node e, unsigned i ) {
+  e = getCandidateForEnumerator( e );
+  std::map< Node, std::vector< Node > >::iterator it = d_examples_out.find( e );
+  if( it!=d_examples_out.end() ){
+    Assert( i<it->second.size() );
+    return it->second[i];
+  }else{
+    Assert( false );
+    return Node::null();
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/ce_guided_pbe.cpp]: This (and above) are what I would call "recoverable" errors, in that if this goes wrong in production, I don't want CVC4 to abort. However, I do want them to give an assertion error in debug.  As I recall, Unreachable() aborts in both debug and production.

File: src/theory/quantifiers/ce_guided_pbe.h (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/ce_guided_pbe.h]: I'd prefer keeping the commits small. I can open a separate issue/PR to clean up this class?

File: src/theory/datatypes/datatypes_sygus.h [hunk 6]
```diff
@@ -33,6 +33,11 @@
 
 namespace CVC4 {
 namespace theory {
+
+namespace quantifiers {
+  class CegConjecture;
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/datatypes/datatypes_sygus.h]: I suppose its fine, I tend to be paranoid about circular includes.

File: src/theory/quantifiers/ce_guided_pbe.h (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/ce_guided_pbe.h]: I've opened #1250.

**Comment by ajreynol** [comment]
I've addressed the review in my latest commit and opened github issue #1250 to clean the Sygus programming-by-examples class.

**Pull request merged**